### DISCUSSION TITLE
How to handle morphing while keeping things DRY with Livewire

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#2147] [1. Help]
I'm working on a project where we have `quotes` and `projects`. 

- Quotes are specific and about one possible project only. The type of project can be different, for example, it might be the development of a specific paint or for the design of a wooden box. You would need to store different details depending on what you are offering.
- Projects would have the same detail as a quote, but it would include all the detail that we gathered during the quotation as a copy from the quote, so projects have their own data, they do not morph to the same entry.

We do this by adding a `detail` property to our `Project` and `Quote` models. For example:

```
public function detail()
{
    return $this->morphOne($this->detailable_by, 'detailable');
}
```

This is all pretty standard in Laravel. Nothing fancy. We would be morphing to the `Paint` and `WoodenBox` models for example. Once a quote is accepted, we make a copy and link it up to the project, that way, if there turns out to be a small change of requirements, we can adjust the project without changing the quote.

Now here's where I'm lost..

A user can edit the details of `Paint` in both the quote as the project. Before Livewire I would probably make either a component or an include for all types of projects (paint, wooden box, lamp, etc.) and have all of the names match up.

With Livewire I have no idea how to approach this case.

They need to link up to a public property, I guess the name isn't important. This could be `$detail`, as it long as it's the same.

- If I have a component or an include, I can wire:model everything to detail.attribute_name (that would work)
- Switching the type of a quote would dynamically reload the component or view (that would also work)
- When switching the type I would probably need to find a way to get only those rules added to the Livewire component that are relevant to the type (no idea how yet)
- When saving I need to remove all unlinked types from every possible type there is to ensure there are no ghost entries, or I should do this as a job that runs every day or so.

Did I forget anything? Do you have any tips or tricks I didn't think about?

**Comment by joshhanley** [comment]
@ju5t I don't completely follow, this confused me:
> A user can edit the details of Paint in both the quote as the project 

Is that meant to be 
>both the quote and the project

To run through a use case, lets say you have a Livewire Quote component, as you have a new quote open on the screen and you want to fill in it's details, you can select the detail type (is that correct) and the fields/types/validation will change depending on which type was selected. And after it's saved the form should be cleared and the detail type fields are removed?

Do I have that right?

If that's the case, if you have a fixed number of types you could create a component for each type, that contains it's fields and validation, and conditionally add the correct component depending on which detail type was selected on the quote component.

Or you could get creative with the validation rules. Instead of using the rules property, you could use a method called rules `protected function rules() {}` and return an array of rules from that method, which you can conditionally change depending on detail type.

Hope this helps!

**Comment by ju5t** [comment]
@joshhanley sorry, I'm not a native speaker. Sometimes I mess up.

You've got the use case right.

I often don't use a component in a component. It means you have to deal with the communication between these components. The logic for that can grow pretty quick. We have the following requirements on how to deal with failure:

- You don't want to save the child component when the parent fails validation.
- You don't want to save the parent component when the child component fails validation.

This means we need the following events and methods:

- A method that starts the validation, this emits an event to the child component.
- The child needs to emit an event passing along its error state. Not sure how to deal with that. Try/catch possibly?
- When the error state comes back to the parent, it needs to validate itself. If it all fails, display all errors. If it's all good, save and do whatever is required after that.

It might be worth the effort though. I'm not sure. There are some benefits too:

- Clear separation. Everything lives in a single component. I like that.
- No need to get fancy with `rules()`. Which I would otherwise have to put somewhere too.

I knew about `rules()`, but I didn't know you could dynamically load a different rule set each time you run `validate()`. I thought it was somehow fixed when the component gets mounted.

I have to think about what would be a good solution here.
